tool/tctl/sso/tester/command.go

```diff
@@ -0,0 +1,266 @@
+		}
+
+		info, infoErr := c.GetSSODiagnosticInfo(context.TODO(), raw.Kind, requestInfo.RequestID)
+
+		err = cmd.reportLoginResult(raw.Kind, info, infoErr, loginResponse, loginErr)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	}
+}
+
+// TryRun is executed after the CLI parsing is done. The command must
+// determine if selectedCommand belongs to it and return match=true
+func (cmd *SSOTestCommand) TryRun(selectedCommand string, c auth.ClientI) (match bool, err error) {
+	if selectedCommand == cmd.ssoTestCmd.FullCommand() {
+		return true, cmd.ssoTestCommand(c)
+	}
+	return false, nil
+}
+
+type AuthRequestInfo struct {
+	Config           *client.RedirectorConfig
+	RequestID        string
+	RequestCreateErr error
+}
+
+func (cmd *SSOTestCommand) runSSOLoginFlow(ctx context.Context, protocol string, c auth.ClientI, config *client.RedirectorConfig) (*auth.SSHLoginResponse, error) {
+	key, err := client.NewKey()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	proxies, err := c.GetProxies()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if len(proxies) == 0 {
+		return nil, trace.BadParameter("cluster has no proxies.")
+	}
+
+	cfg := client.MakeDefaultConfig()
+	cfg.WebProxyAddr = proxies[0].GetPublicAddr()
+
+	tc, err := client.NewClient(cfg)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return client.SSHAgentSSOLogin(ctx, client.SSHLoginSSO{
+		SSHLogin: client.SSHLogin{
+			ProxyAddr:         tc.WebProxyAddr,
+			PubKey:            key.Pub,
+			TTL:               tc.KeyTTL,
+			Insecure:          tc.InsecureSkipVerify,
+			Pool:              nil,
+			Compatibility:     tc.CertificateFormat,
+			RouteToCluster:    tc.SiteName,
+			KubernetesCluster: tc.KubernetesCluster,
+		},
+		ConnectorID: "-sso-test",
+		Protocol:    protocol,
+		BindAddr:    tc.BindAddr,
+		Browser:     tc.Browser,
+	}, config)
+}
+
+type DiagnosticInfoField struct {
+	Present bool
+	Show    bool
+	Msg     string
+}
+
+func (cmd *SSOTestCommand) reportLoginResult(authKind string, diag *types.SSODiagnosticInfo, infoErr error, loginResponse *auth.SSHLoginResponse, loginErr error) (errResult error) {
+	success := diag != nil && diag.Success
+
+	// check for errors
+	if loginErr != nil || infoErr != nil {
+		success = false
+	}
+
+	if success {
+		fmt.Printf("Success! Logged in as: %v\n", loginResponse.Username)
+	} else {
+		fmt.Printf("Failure!\n")
+		errResult = trace.Errorf("SSO flow failed.")
+
+		if infoErr != nil {
+			fmt.Printf("No diagnostic info found. Most likely cause: the request timed out or callback configuration is incorrect. Ensure that user logs within alloted time and IdP configuration is correct.\n Error details: %v\n", trace.UserMessage(infoErr))
+			errResult = trace.Wrap(infoErr, "SSO flow failed.")
+		}
+
+		if loginErr != nil {
+			fmt.Printf("Login error: %v\n", trace.UserMessage(loginErr))
+			errResult = trace.Wrap(loginErr, "SSO flow failed.")
+		}
```

Comment:
what if `infoErr != nil` and  `loginErr != nil` in that case the `loginErr` take precedence over `infoErr` 
Can we use `trace.Aggragate(infoErr, loginErr)` in this case ?

Reply:
We could, but this is deliberate: if both are non-nil, then `loginErr` will hold the important stuff, while `infoErr` will be just noise.